background/index.ts

```diff
@@ -94,6 +97,172 @@ async function main() {
+                console.log(
+                  `[background][indexBalanceOnBlock] ERC20 token ${
+                    t.symbol
+                  } balance updated: ${tokenBalance.toString()} at chain ID: ${chainId}`
+                )
+              }
+            })()
+          )
+        })
+      })
+
+      // Execute all balance queries
+      try {
+        await Promise.all(tokenQueries)
+      } catch (error) {
+        console.error(
+          `[background][indexBalanceOnBlock] Error executing token balance updates: ${error}`
+        )
+      }
+    }
+
+    // Handle accountActive state changes and bind providers as needed
+    const networkActiveStateSubscriber = async () => {
+      // Function to handle block updates using the provider context
+      const blockSubscriberWithProvider = async function (
+        this: { provider: JsonRpcProvider; chainId: number },
+        _: number
+      ) {
+        // Get the latest accounts
+        const { account } = storage.get()
+
+        const accountsForThisNetwork = Object.values(account).filter(
+          (account) => account.chainId === this.chainId
+        )
+        if (accountsForThisNetwork.length <= 0) {
+          return
+        }
+
+        await updateAccountBalances(accountsForThisNetwork, this.provider)
+      }
+
+      const { network, networkActive } = storage.get()
+      const networkInstance = network[networkActive]
+      const { accountActive, nodeRpcUrl, chainId } = networkInstance
+
+      // Set `{ staticNetwork: true }` to avoid infinite retries if nodeRpcUrl fails.
+      // Refer: https://github.com/ethers-io/ethers.js/issues/4377
+      const provider = new JsonRpcProvider(nodeRpcUrl, chainId, {
+        staticNetwork: true
+      })
+
+      const blockSubscriber = blockSubscriberWithProvider.bind({
+        provider,
+        chainId
+      })
+
+      // Update provider and subscriber on network switch
+      if (
+        blockSubscriberContext.provider &&
+        blockSubscriberContext.blockSubscriber
+      ) {
+        blockSubscriberContext.provider.off(
+          "block",
+          blockSubscriberContext.blockSubscriber
+        )
+      }
+
+      blockSubscriberContext.provider = provider
+      blockSubscriberContext.blockSubscriber = blockSubscriber
+
+      // Listen for new blocks
+      blockSubscriberContext.provider.on("block", blockSubscriber)
+      console.log(
+        `[background][indexBalanceOnBlock] Provider listening for blocks on chainId: ${chainId}`
+      )
+
+      if (!accountActive) {
+        return
+      }
+
+      const { account } = storage.get()
+      const accounts = Object.values(account)
+
+      // Group accounts by chain ID
+      const accountsByNetwork = accounts.filter(
+        (account) => account.chainId === chainId
+      )
+
+      // First update balances for the local testnet makeup.
+      await updateAccountBalances(
+        accountsByNetwork,
+        blockSubscriberContext.provider
+      )
```

Comment:
Here are almost same as `blockSubscriber`, what about calling `blockSubscriber` directly instead?

Reply:
I have already modified it in Commit [https://github.com/pilagod/waallet-extension/pull/202/commits/785ccc77f55527206ccbdc61c32768b12b9e3acd](https://github.com/pilagod/waallet-extension/pull/202/commits/785ccc77f55527206ccbdc61c32768b12b9e3acd).
I found that the listener can catch an async void return function and still listen and execute normally.